Firestore/core/test/firebase/firestore/local/remote_document_cache_test.mm

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/test/firebase/firestore/local/remote_document_cache_test.h"
+
+#include <memory>
+#include <vector>
+
+#include "Firestore/core/src/firebase/firestore/local/memory_remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/local/persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/remote_document_cache.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+#include "Firestore/core/src/firebase/firestore/util/string_apple.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "absl/strings/string_view.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using model::Document;
+using model::DocumentKey;
+using model::DocumentKeySet;
+using model::DocumentMap;
+using model::DocumentState;
+using model::FieldValue;
+using model::MaybeDocument;
+using model::MaybeDocumentMap;
+using model::NoDocument;
+using model::OptionalMaybeDocumentMap;
+
+using testing::Matches;
+using testing::IsSupersetOf;
+using testing::UnorderedElementsAreArray;
+using testutil::DeletedDoc;
+using testutil::Doc;
+using testutil::Map;
+using testutil::Query;
+
+const char* kDocPath = "a/b";
+const char* kLongDocPath = "a/b/c/d/e/f";
+const int kVersion = 42;
+FieldValue::Map kDocData;
+
+template <typename MapType>
+std::vector<MaybeDocument> ActualDocuments(const MapType& docs) {
+  std::vector<MaybeDocument> result;
```

Comment:
Neither the name nor the function signature convey a whole lot about what this method does. Would it be possible to not use a template here (the function always seems to convert from an OptionalMaybeDocumentMap to `std::vector<MaybeDocument>`) and to use a more descriptive name (`ExtractDocuments` or `ExtractExistingDocuments`)?

Reply:
Somehow I had convinced myself the template was required but clearly it isn't. Changed to `OptionalMaybeDocumentMap`.

Good idea re the name.

Reply:
The update here is causing a build break for me. I temporarily reverted it out in my branch.

Reply:
Yeah, it turns out that the template was required because we do this for both `OptionalMaybeDocumentMap` and `MaybeDocumentMap`.

The reason this code works in both cases is that `optional<MaybeDocument>` will implicitly convert from `MaybeDocument`. This means that in the case of the `MaybeDocumentMap` the assignment to the optional creates a temporary that has a guaranteed value and the if statement is always true.

Another way to resolve this would be to provide overloads for both map types, but this seems annoying.